### DISCUSSION TITLE
Is a NACK rejection a transmission error?

### DISCUSSION THREAD

**Discussion by gbhoffman** [nextgenhealthcare/connect#6066] [Q&A] [answered]
Hello,

Using MirthConnect 4.2.0 with hundreds of successful transmissions a day to an upstream OnBase gateway, I'm starting to get NACK error messages such as the following:

> NACK sent from receiver. (Rejected)
> HL7 Inbound processor [Create Advanced Doc]: Error writing message file to disk group: C:\E\Users\E\OBLIST\R\1\E\AppData\E\Local\E\Temp\E\OB4907.pdf
>  
> MSH|^~\&|EPICADT||PROVIDERS|ONBASE|20231130062803||ACK|199|P|2.4||||||||
> MSA|AR||HL7 Inbound processor [Create Advanced Doc]: Error writing message file to disk group: C:\E\Users\E\OBLIST\R\1\E\AppData\E\Local\E\Temp\E\OB4907.pdf||||

There's obviously something wrong on the receiving end, which is someone else's problem.

I've set up the channel to retry upon error, but these messages are not being retried and are still shown as "processing." I guess Mirth doesn't take rejection lightly.

Any suggestions on how I can set up the channel to either retry the transmission or report the error to me?

Thanks,
Gary

**Comment by jonbartels** [comment, marked as answer]
> Any suggestions on how I can set up the channel to either retry the transmission or report the error to me?

I think you have two options:
1. Use [response validation](https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_Destination_Settings_connect_ug.html) and [datatype settings](https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_HL7_v2_x_Data_Type_connect_ug.html) - set up a response validation to mark the AR types as ERROR. I haven't done this in a while but I think you might just need to turn on response validation. AR is already a default error code.
2. Write your own Response transformer. In the response transformer `msg` will be the ACK/NACK/error you got back. You can check that message for certain NACK codes or error messages. Then you force the message state to QUEUED or ERRORED.

Once either of those options gets your message to a QUEUED or ERROR state you can resend it or get an alert or whatever.

**Comment by gbhoffman** [comment]
In the Destination Settings:
Queue Messages is set to On Failure
Advanced Queue Settings sets 5 retries with an interfal of 10000 ms.
Validate Response is Yes. 

The messages with the AR ACK code do go to Error status. However, they apparently don't reprocess/resend.

If I reprocess the errored messages by hand (right click), they will send and so far, they don't error. (The problem with the downstream system is random and intermittent.)

Do is there another setting I'm missing to automatically reprocess the outbound messages?

Thanks,
Gary

**Reply by jonbartels** [reply]
Got it, so you want option 2 from above sort of.

Check the message status and the send attempts count. Then you can force the message back to the queued status and let mirth try again

**Reply by jonbartels** [reply]
Posting a followup with a code sample. I have not tested this but it should be close.

In your response transformer:
```
var sendAttemptsLimit = 5; //TODO you should make this a configmap value
//if the message errored AND we have tried less than the limit, force back to the queue
if(responseStatus == ERROR && connectorMessage.getSendAttempts() <= sendAttemptsLimit ){
   //TODO optional - you can add some logic here and look at msg['MSA'] or msg['ERR'] and read the HL7 NACK reason
   //TODO - reading the reason would let you decide that like a patient not found gets re-tried but some other error is really an error and should fail

    //put a connector map value to show why we retried, this is better than a log because it stays with the message 
    $co('resendReason', 'Got error but have tried less than ' + sendAttemptsLimit + ' times. Re-queueueueueing');
    responseStatus = QUEUED
}
```

> Advanced Queue Settings sets 5 retries with an interfal of 10000 ms.

This is good. It means you will try once every 10 seconds. 

There is some additional tuning that can be done with threads, rotate queue, and the thread assignment variable. This WILL be sensitive to message order and the delay you need before re-sending however.
